File: leetcode/problem_5.py
# Problem 5: Longest Palindromic Substring (Medium): https://leetcode.com/problems/longest-palindromic-substring/

class Solution:
    def longestPalindrome(self, s: str) -> str:
        maxLen = 0
        result =''
        if(len(s) < 2):
            return s
        for i in range(len(s)-1):
            maxPal1 = self.maxPalindromeLenAtIndex(i,i, s)
            maxPal2 = self.maxPalindromeLenAtIndex(i, i+1,s)
            maxPal = ''
            if(len(maxPal1) > len(maxPal2)):
                maxPal = maxPal1
            else:
                maxPal = maxPal2
            if(len(maxPal) > maxLen):
                maxLen = len(maxPal)
                result = maxPal
        return result

    def maxPalindromeLenAtIndex(self, b:int, e:int, s: str) -> str:
        maxLen = 0
        start = b
        end = e
        maxPal = ''
        while(start >= 0 and end < len(s) and s[start] == s[end]):
            strLen = end - start + 1
            subStr = s[start:end+1]
            if(strLen > maxLen):
                maxLen = strLen
                maxPal = subStr
            start -= 1
            end += 1
        return maxPal

# for each char in String expand outward
# same logic for even and odd string has to be repeated
class Solution2:
    def longestPalindrome(self, s: str) -> str:
        result = ""
        for i in range(len(s)):
            start = i
            end = i
            while(start >= 0 and end < len(s) and s[start] == s[end]):
                if(len(result) < end - start + 1):
                    result = s[start : end+1]
                start -=1
                end += 1
            start = i
            end = i+1
            while(start >= 0 and end < len(s) and s[start] == s[end]):
                if(len(result) < end - start + 1):
                    result = s[start : end + 1]
                start -= 1
                end += 1
        return result


# Brute Force

class Solution3:
    def longestPalindrome(self, s: str) -> str:
        def isPalindrome(s):
            return s == s[:: -1]
        if(len(s) < 2):
            return s
        max_len = 1
        result = s[0]
        for i in range(len(s)-1):
            for j in range(i+1, len(s)):
                if(j - i + 1 > max_len and isPalindrome(s[i:j+1])):
                    result = s[i:j+1]
                    max_len = j - i + 1
        return result


if __name__ == "__main__":
        s = Solution()
        print(s.longestPalindrome('ababcccbabb'))

File: leetcode/problem_5.kt
// Problem 5. Longest Palindromic Substring
// tags: String, Two pointer Dynamic Programming

// time complexity: O(n^3)
// expands right
class Solution {
    fun longestPalindrome(s: String): String {
        /* given string
         * output: longest palindromic string
         * abcbasaabbcbbabcbbcbbaababac
         *
         */
        var result: String = ""
        var length = 0
        for (i in 0..s.length - 1) {
            for (j in i..s.length - 1) {
                var start = i
                var end = j
                var isPalindrome: Boolean = true
                while (start <= end) {
                    if (s[start] == s[end]) {
                        start += 1
                        end -= 1
                    } else {
                        isPalindrome = false
                        break
                    }
                }
                if (isPalindrome) {
                    if (j - i + 1 > length) {
                        result = s.substring(i, j + 1)
                        length = j - i + 1
                    }
                }
            }
        }
        return result
    }
}

// Slightly Optimized
// compresses the string
// time complexity: O(n^3)
class Solution {
    fun longestPalindrome(s: String): String {
        /* given string
         * output: longest palindromic string
         * abcbasaabbcbbabcbbcbbaababac
         *
         */
        var result: String = ""
        var length = 0
        for (i in 0..s.length - 1) {
            for (j in s.length - 1 downTo 0 step 1) {
                var start = i
                var end = j
                var isPalindrome: Boolean = true
                while (start <= end) {
                    if (s[start] == s[end]) {
                        start += 1
                        end -= 1
                    } else {
                        isPalindrome = false
                        break
                    }
                }
                if (isPalindrome) {
                    if (j - i + 1 > length) {
                        result = s.substring(i, j + 1)
                        length = j - i + 1
                    }
                    break
                }
            }
        }
        return result
    }
}


// Optimized solution
// time complexity: O(n^2)
// expand outward at each index to check if its a palindrome
// same as problem 647
class SolutionOptimized {
    fun longestPalindrome(s: String): String {
        /* given string
         * output: longest palindromic string
         * abcbasaabbcbbabcbbcbbaababac
         *
        */
        var result = ""
        var length = 0
        for(i in s.indices){
            var start = i
            var end = i
            var oddResult = longestPalindromicString(start, end, result, s)
            // even length palindrome
            start = i
            end = start + 1
            result = longestPalindromicString(start, end, oddResult, s)
        }
        return result
    }

    fun longestPalindromicString(i:Int, j: Int, res:String, s:String):String{
        var length = res.length
        var result = res
        var start = i
        var end = j
        while(start >= 0 && end < s.length && s[start] == s[end]){
            if(end - start + 1 > length){
                result = s.substring(start, end + 1)
                length = end - start + 1
            }
            start -= 1
            end += 1
        }
        return result
    }
}